### ISSUE TITLE
IDE0078 - Use pattern matching removes necessary side-effects in some cases

### ISSUE THREAD

**Issue by bartdesmet** [dotnet/roslyn-analyzers#4690]
Consider the following fragment from some parser using a `reader` to get characters:

```csharp
                case 'n':
                    // null
                    {
                        reader.Read();
                        if (reader.Read() != 'u' || reader.Read() != 'l' || reader.Read() != 'l')
                            throw new ParseException(/* omitted details */);
                    }
                    break;
```

IDE0078 suggests to rewrite this as:

```csharp
                case 'n':
                    // null
                    {
                        reader.Read();
                        if (reader.Read() is not 'u' or not l' or not 'l')
                            throw new ParseException(/* omitted details */);
                    }
                    break;
```

which removes the `Read()` side-effect, and is immediately followed by another diagnostic that indicates that the pattern always matches.

Some more constraints may have to be enforced on the common expression (here `reader.Read()`) to filter out false positives where this rewrite is unsafe. Ideally, the expression is "pure" but that may be too stringent (e.g. property getters would have to be considered as potentially impure).

**Comment by bartdesmet** [comment]
@mavasani this may be in the wrong repo as well. Can you check if this belongs in the roslyn repo instead?
